### ISSUE TITLE
PAM authentication issues on Debian

### ISSUE THREAD

**Issue by laurikoobas** [jupyterhub/jupyterhub#1026]
Seems to be broken similar to some older closed issues.

The issue: PAM authentication fails.

The longer version of the issue - the exact same code for the exact same (test) user works when called with python3, but does not work when called by jupyterhub (0.7.2).

I have a test user - testuser.
This is the code that works:
```python
import pamela

username = 'testuser'
password = 'testuser'
service = 'login'

try:
    print("PAM Authentication test. username = {0} and service = {1}".format(username, service))
    pamela.authenticate(username, password, service)
except pamela.PAMError as e:
    print("PAM Authentication failed: {0}".format(e))
else:
    print("success!")
```

It outputs:
```
jupyterhub@virtualmachine:/tmp$ python3 pamtest.py
PAM Authentication test. username = testuser and service = login
success!
```

And this is the code that does not work, found (and sligthly modified) in:
/usr/local/lib/python3.4/dist-packages/jupyterhub/auth.py
```python
    def authenticate(self, handler, data):
        """Authenticate with PAM, and return the username if login is successful.

        Return None otherwise.
        """
        username = data['username']
        import traceback
        traceback.print_stack()
        try:
            self.log.warning("PAM Authentication test. username = %s and service = %s", username, self.service)
            # pamela.authenticate(username, data['password'], service=self.service)
            pamela.authenticate('testuser', 'testuser', 'login')
        except pamela.PAMError as e:
            if handler is not None:
                self.log.warning("PAM Authentication failed (%s@%s): %s", username, handler.request.remote_ip, e)
            else:
                self.log.warning("PAM Authentication failed: %s", e)
        else:
            return username
```

This is the output:
```
[I 2017-03-17 10:32:58.645 JupyterHub app:1485] JupyterHub is now running at http://127.0.0.1:8000/jupyter/
  File "/usr/local/bin/jupyterhub", line 4, in <module>
    main()
  File "/usr/local/lib/python3.4/dist-packages/jupyterhub/app.py", line 1536, in launch_instance
    loop.start()
  File "/usr/local/lib/python3.4/dist-packages/tornado/ioloop.py", line 831, in start
    self._run_callback(callback)
  File "/usr/local/lib/python3.4/dist-packages/tornado/ioloop.py", line 604, in _run_callback
    ret = callback()
  File "/usr/local/lib/python3.4/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 1104, in <lambda>
    self.future, lambda f: self.run())
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 1024, in run
    yielded = self.gen.send(value)
  File "/usr/local/lib/python3.4/dist-packages/tornado/http1connection.py", line 238, in _read_message
    delegate.finish()
  File "/usr/local/lib/python3.4/dist-packages/tornado/httpserver.py", line 289, in finish
    self.delegate.finish()
  File "/usr/local/lib/python3.4/dist-packages/tornado/web.py", line 2047, in finish
    self.execute()
  File "/usr/local/lib/python3.4/dist-packages/tornado/web.py", line 2080, in execute
    **self.path_kwargs)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 285, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python3.4/dist-packages/tornado/web.py", line 1467, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 285, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python3.4/dist-packages/jupyterhub/handlers/login.py", line 72, in post
    username = yield self.authenticate(data)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 285, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python3.4/dist-packages/jupyterhub/handlers/base.py", line 281, in authenticate
    result = yield auth.get_authenticated_user(self, data)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 285, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python3.4/dist-packages/jupyterhub/auth.py", line 171, in get_authenticated_user
    username = yield self.authenticate(handler, data)
  File "/usr/local/lib/python3.4/dist-packages/tornado/gen.py", line 270, in wrapper
    result = func(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/jupyterhub/auth.py", line 470, in authenticate
    traceback.print_stack()
[W 2017-03-17 10:33:06.312 JupyterHub auth:472] PAM Authentication test. username = testuser and service = login
[W 2017-03-17 10:33:09.253 JupyterHub auth:477] PAM Authentication failed (testuser@127.0.0.1): [PAM Error 7] Authentication failure
[D 2017-03-17 10:33:09.253 JupyterHub login:95] Failed login for testuser
```

Something fails about how jupyterhub calls it.
The /var/log/auth.log shows the following:
```
Mar 17 03:48:08 virtualmachine unix_chkpwd[35948]: check pass; user unknown
Mar 17 03:48:08 virtualmachine unix_chkpwd[35948]: password check failed for user (testuser)
Mar 17 03:48:08 virtualmachine python3: pam_unix(login:auth): authentication failure; logname=something uid=1006 euid=1006 tty= ruser= rhost=  user=testuser
```

I looked into getting the PAM module to output some debug data, but couldn't figure out how to do it.

**Comment by minrk** [comment]
Your successful test output has `username = kaalikas`, not testuser. Are you running the Hub as kaalikas? Does login to the Hub work as kaalikas?

**Comment by laurikoobas** [comment]
Damit, well.. my testuser is named "kaalikas" and I wanted to replace it everywhere and failed :)

I'm running the hub as user "jupyterhub". Logging in with "jupyterhub" works actually. But any other user, either in admin list or not, does not work.

**Comment by minrk** [comment]
No problem,the situation is probably what I was trying to get at with that question. When the Hub can authenticate the user it is running as but not anyone else, this generally means that the Hub user (`jupyterhub`) doesn't have permission to check other users' passwords. On basic Linux systems, this means read permissions on the shadow password database (`/etc/shadow`). On Ubuntu, this file is owned by the `shadow` group, so giving `jupyterhub` permission to read it is adding that user to the group:

    sudo usermod -a -G shadow jupyterhub

The wiki has [more details](https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges#enabling-pam-for-non-root) on this.

If you run your pamtest, make sure to run it *as jupyterhub* to test the permissions, e.g.

    sudo -u jupyterhub python3 pamtest.py

and make sure that the username you are testing is not the username running the test, because users are always able to check their own password.

**Comment by laurikoobas** [comment]
All of that works. I guess I should have mentioned that I went through that guide to begin with.
```
lauri@virtualmachine:/tmp$ less /etc/group | grep shadow
shadow:x:42:jupyterhub
```
```
lauri@virtualmachine:/tmp$ sudo -u jupyterhub python3 /tmp/pamtest.py
[sudo] password for lauri:
PAM Authentication test. username = testuser and service = login
success!
```

**Comment by minrk** [comment]
Hm, so *as jupyterhub* the test script works, then as the same user jupyterhub doesn't? Does it matter which user attempts to login first? Are you running jupyterhub in the same terminal session where you are running the test, or via systemd/supervisor/etc.?

**Comment by laurikoobas** [comment]
> so as jupyterhub the test script works, then as the same user jupyterhub doesn't?

Seems to be the case, but I may have more information below.

> Does it matter which user attempts to login first?

nope

> Are you running jupyterhub in the same terminal session where you are running the test, or via systemd/supervisor/etc.?

Was running everything in terminal session.

Tried a few more things, just trying to be thorough.
First I tried the PAM test from the guide. Entered the correct password first and then incorrect one.
```
lauri@virtualmachine:~$ sudo -u jupyterhub python3 -c "import pamela, getpass; print(pamela.authenticate('$USER', getpass.getpass()))"
[sudo] password for lauri:
Password:
None
lauri@virtualmachine:~$ sudo -u jupyterhub python3 -c "import pamela, getpass; print(pamela.authenticate('$USER', getpass.getpass()))"
Password:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/pamela.py", line 285, in authenticate
    return pam_end(handle, retval)
  File "/usr/local/lib/python3.4/dist-packages/pamela.py", line 244, in pam_end
    raise PAMError(errno=retval)
pamela.PAMError: [PAM Error 7] Authentication failure
```
So I guess that works as expected?

Then I ran JupyterHub through sudo like this:
```
sudo -u jupyterhub jupyterhub -f /etc/jupyterhub/jupyterhub_config.py --debug
```
And tried to log in with jupyterhub (*with c.PAMAuthenticator.open_sessions = False*)
```
[I 2017-03-17 14:43:08.129 JupyterHub log:100] 200 GET /jupyter/hub/login (@127.0.0.1) 5.97ms
[W 2017-03-17 14:43:09.928 JupyterHub auth:472] PAM Authentication test. username = jupyterhub and service = login
[I 170317 14:43:10 mediator:75] Spawning /usr/local/bin/jupyterhub-singleuser '--user="jupyterhub"' '--cookie-name="jupyter-hub-token-jupyterhub"' '--base-url="/jupyter/user/jupyterhub"' '--hub-host=""' '--hub-prefix="/jupyter/hub/"' '--hub-api-url="http://127.0.0.1:8081/jupyter/hub/api"' '--ip="127.0.0.1"' --port=55075
[W 2017-03-17 14:43:10.828 jupyterhub notebookapp:256] Widgets are unavailable. Please install widgetsnbextension or ipywidgets 4.0
[W 2017-03-17 14:43:10.835 jupyterhub login:222] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2017-03-17 14:43:10.840 jupyterhub notebookapp:1201] Serving notebooks from local directory: /home/jupyterhub
[I 2017-03-17 14:43:10.840 jupyterhub notebookapp:1201] 0 active kernels
[I 2017-03-17 14:43:10.841 jupyterhub notebookapp:1201] The Jupyter Notebook is running at: http://127.0.0.1:55075/jupyter/user/jupyterhub/
[I 2017-03-17 14:43:10.841 jupyterhub notebookapp:1202] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[D 2017-03-17 14:43:11.127 JupyterHub spawner:584] Polling subprocess every 30s
[I 2017-03-17 14:43:11.139 jupyterhub log:47] 302 GET /jupyter/user/jupyterhub (127.0.0.1) 0.57ms
[D 2017-03-17 14:43:11.140 JupyterHub utils:84] Server at http://127.0.0.1:55075/jupyter/user/jupyterhub responded with 302
[I 2017-03-17 14:43:11.140 JupyterHub base:322] User jupyterhub server took 1.202 seconds to start
[I 2017-03-17 14:43:11.140 JupyterHub orm:188] Adding user jupyterhub to proxy /jupyter/user/jupyterhub => http://127.0.0.1:55075
[D 2017-03-17 14:43:11.142 JupyterHub orm:146] Fetching POST http://127.0.0.1:8001/api/routes/jupyter/user/jupyterhub
[D 2017-03-17 14:43:11.145 JupyterHub base:236] Setting cookie for jupyterhub: jupyter-hub-token-jupyterhub, {}
[D 2017-03-17 14:43:11.148 JupyterHub base:236] Setting cookie for jupyterhub: jupyter-hub-token, {}
[I 2017-03-17 14:43:11.151 JupyterHub log:100] 302 POST /jupyter/hub/login?next= (@127.0.0.1) 1221.47ms
[I 2017-03-17 14:43:11.151 JupyterHub login:91] User logged in: jupyterhub
```
And it's a success! I can see the admin panel and whatnot.

Then tried with my test user:
```
[W 2017-03-17 14:48:56.032 JupyterHub auth:472] PAM Authentication test. username = testuser and service = login
sudo: a password is required
[E 2017-03-17 14:48:56.072 JupyterHub user:251] Unhandled error starting testuser's server: substring not found
[E 2017-03-17 14:48:56.085 JupyterHub web:1548] Uncaught exception POST /jupyter/hub/login?next= (127.0.0.1)
    HTTPServerRequest(protocol='http', host='blaah', method='POST', uri='/jupyter/hub/login?next=', version='HTTP/1.1', remote_ip='127.0.0.1', headers={blaah}})
    Traceback (most recent call last):
      File "/usr/local/lib/python3.4/dist-packages/tornado/web.py", line 1469, in _execute
        result = yield result
      File "/usr/local/lib/python3.4/dist-packages/jupyterhub/handlers/login.py", line 84, in post
        yield self.spawn_single_user(user)
      File "/usr/local/lib/python3.4/dist-packages/jupyterhub/handlers/base.py", line 328, in spawn_single_user
        yield gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), f)
      File "/usr/local/lib/python3.4/dist-packages/jupyterhub/user.py", line 261, in spawn
        raise e
      File "/usr/local/lib/python3.4/dist-packages/jupyterhub/user.py", line 229, in spawn
        ip_port = yield gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
      File "/usr/local/lib/python3.4/dist-packages/sudospawner/spawner.py", line 83, in start
        reply = yield self.do(action='spawn', args=self.get_args(), env=self.get_env())
      File "/usr/local/lib/python3.4/dist-packages/sudospawner/spawner.py", line 69, in do
        data_str = data_str[data_str.index('{'):data_str.rindex('}')+1]
    ValueError: substring not found

[D 2017-03-17 14:48:56.087 JupyterHub base:462] No template for 500
[E 2017-03-17 14:48:56.090 JupyterHub log:99] {
      "X-Forwarded-Port": "80",
      .....
      "Connection": "close",
    }
[E 2017-03-17 14:48:56.091 JupyterHub log:100] 500 POST /jupyter/hub/login?next= (@127.0.0.1) 57.35ms

``` 

Then I re-did the same login trials with *c.PAMAuthenticator.open_sessions = TRUE*.
The results were the same (jupyterhub success, testuser failure), but some extra information appeared in BOTH cases the log like shown below:
```
[W 2017-03-17 14:53:12.664 JupyterHub auth:472] PAM Authentication test. username = testuser and service = login
Linux virtualmachine 3.16.0-4-amd64 #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) x86_64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
[W 2017-03-17 14:53:12.706 JupyterHub auth:490] Failed to open PAM session for testuser: [PAM Error 14] Cannot make/remove an entry for the specified session
[W 2017-03-17 14:53:12.706 JupyterHub auth:491] Disabling PAM sessions from now on.
sudo: a password is required
[E 2017-03-17 14:53:12.723 JupyterHub user:251] Unhandled error starting testuser's server: substring not found
```

**Comment by minrk** [comment]
The substring not found suggests that perhaps testuser can't start its single-user server. Can you try:

    sudo -u testuser jupyterhub-singleuser

?

**Comment by laurikoobas** [comment]
Well, that just runs into a warning / error I don't know how to bypass:
```
JUPYTERHUB_API_TOKEN env is required to run jupyterhub-singleuser. Did you launch it manually?
```

**Comment by laurikoobas** [comment]
Well, I got it working finally after very very verbose logging and a lightbulb going off last night.

The [document ](https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges) about running sudospawner without root. It has the line about "Runas_Alias JUPYTER_USERS = zoe, wash". I went with the comments above it ("_Give the new user ..._") and assumed that it meant the "rhea" user, but the sudoers part under it specifies some other users like "zoe" and "wash" there as well.

Apparently it means that the list there (JUPYTER_USERS) needs to contain everyone that should be able to run/start/access/use their own notebook server, not just the user that runs the hub. I guess the wording could be better :) Especially as it's the hub process that initiates the whole thing.

My sudo setup test that I showed here previously succeeded due to the user (my actual user that I use to log in to the virtual machine for administrative tasks) is obviously already in sudo group. And hence could run the sudospawner as well through the $USER variable carrying on the name.

So, in conclusion, things are working, but really really not obvious.
